### ISSUE TITLE
Theme is computed incorrectly

### ISSUE THREAD

**Issue by denis-anisimov** [vaadin/flow#5762]
`FrontendDependencies::computeEndpoints()` goes through all classes annotated with  `Route`.
Then it has weird logic:

```
if (rootData == null && data.route.isEmpty()) {
                rootData = data;
            }
```
and in the end :

`setTheme(rootData);` 

So the first endpoint data (`data` ) which has route path `""` is used for theming.

Why ?
In which way  `""` is better than any other route ?

As I understand we support only one theme.
So if there are many endpoints data then we should check:
- if there is  `@Theme` annotation at all.
- If there is no `@Theme` annotation then use  `null` theme.
- If there are several different `@Theme` annotations : throw.
- If there is only one `@Theme` annotation then use it (even though there are view without this annotation at all).

**Comment by manolo** [comment]
The implementation follows the documentation, we already discussed this with @bogdanudrescu who changed the initial implementation.
Based on  https://vaadin.com/docs/v13/flow/theme/using-component-themes.html the rules are

1. If a @Theme annotation is found at the root navigation level, the theme defined by it is used;
2. If a @NoTheme annotation is found at the root navigation level, theming is disabled;
3. If the com.vaadin.flow.theme.lumo.Lumo class is available in the classpath (which comes from the vaadin-lumo-theme project), then it is used as the default theme;

So we should ignore any Theme annotation not happening in `@Route(value = "")`

**Comment by manolo** [comment]
> In which way "" is better than any other route ?

`""` means root navigation level, spec in documentation only consider themes at root level, any other should be ignored, it might be thrown but that could cause other problems

**Comment by denis-anisimov** [comment]
Сжалься- the root navigation level

What is the root navigation level ?
That's the main confusion point.

The current implementation considers root navigation level as `@Route("")`.
This is totally wrong.

The root navigation level is:
- _Any_  `@Route`  component which has no top level route layout.
- A route layout component which has no parent route layout.

So this is a root navigation level:
```
@Route("any_string_here_")
public class MyView extends Component {
}
```

_and_ this is also a root navigation level :
```
@Route("")
public class AnotherView extends Component {
}
```

And both of them if they are in the same project.
Theme is applied based on which view currently is active (and not which _may be_ have an annotation and `""`  as a path).

Also this one is root navigation level

```
class MyLayout extends Component implements RouterLayout {
}
```

And this one is not: 

```
@Route("", layout =MyLayout.class)
public class AView extends Component {
}
```

So @Theme can be on `MyLayout` but it's ignored or an error is reported for `AView`.

Please read the algorithm which I've written how theme should be calculated.

Summon @caalador here to confirm.

**Comment by denis-anisimov** [comment]
> > In which way "" is better than any other route ?
> 
> `""` means root navigation level, spec in documentation only consider themes at root level, any other should be ignored, it might be thrown but that could cause other problems

Who said that ?
I don't see any logical reason for this.
And this differs dramatically from how it has been done in V13.

The usecase is _VERY_ simple.

I have two views:

```
@Route("")
class A extends Component {
}

@Route("a")
@Theme(MyTheme.class)
class B extends Component {
}
```

Current impl will use `null`  as a theme which is absolutely wrong: it might be that I don't fo to the `""`  at all.
Both routes `""`  and `"a"` are absolutely equal. It's just matter of which you consider as a primary one and how navigation is set up.

**Comment by denis-anisimov** [comment]
And just one more logical issue: what if I just don't have `@Route("")`  at all but I have a bunch of routes which are all with the same theme with non-empty paths. Is `null` still correct ? Nonsense.

**Comment by manolo** [comment]
So what is the concept of root level, where is it defined? we miss understood that during the implementation then

**Comment by caalador** [comment]
The original theme function is so that during development mode the Theme is dependent on the first loaded `@Route` and could basically be different for different route entrypoints. (This shouldn't be done)

For a production build using Maven all `@Theme` annotations are collected and if there exists 2 different themes an exception is thrown else the theme is selected or lumo if none defined and it exists on the classpath.

So basically now we should consider all `@Route` annotations and get the "single" defined theme and use that or lumo if none defined and found on the classpath.

**Comment by denis-anisimov** [comment]
> So what is the concept of root level, where is it defined? we miss understood that during the implementation then

That's the problem.
Too many confusion.
I've described the root navigation level component above.
I'm not sure about route layouts. @caalador should clarify this. We had a code which calculated a root navigation component based on the current view. Don't know which state it has at the moment.

But what should work definitely is : if a component annotated with `@Route`  and has no parent layout then it's a root navigation component.

**Comment by denis-anisimov** [comment]
> The original theme function is so that during development mode the Theme is dependent on the first loaded `@Route` and could basically be different for different route entrypoints. (This shouldn't be done)
> 
> For a production build using Maven all `@Theme` annotations are collected and if there exists 2 different themes an exception is thrown else the theme is selected or lumo if none defined and it exists on the classpath.
> 
> So basically now we should consider all `@Route` annotations and get the "single" defined theme and use that or lumo if none defined and found on the classpath.

Right. Which is the same as I've written in the ticket description.
Thanks @caalador

**Comment by bogdanudrescu** [comment]
It seems I misunderstood the logic in the documentation. I'll try to fix this. Thanks for noticing it @denis-anisimov

**Comment by manolo** [comment]
It seems also misunderstood the theme features and specs in the implementation for V13, or at least the implementation is not coherent with the documentation

In dev-mode
  - You can have multiple views, and each one with its own `@Theme`
  - You can have views with `@NoTheme` and other with `@Theme` and apparently nothing happens, theme is applied.
  - You can have different variants per view and it works

In prod-mode
  - It's only considered the first class annotated with `@Theme` (does not takes in account Route or Layouts)
  - `@NoTheme` annotation is not considered at all
  -  First class found variant is applied for all the app

**Comment by mehdi-vaadin** [comment]
The JavaDoc for `NoTheme` should also be updated.
